{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "afc0491c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting Bio\n",
      "  Downloading bio-1.5.8-py3-none-any.whl (276 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m276.3/276.3 kB\u001b[0m \u001b[31m1.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: requests in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from Bio) (2.28.1)\n",
      "Collecting gprofiler-official\n",
      "  Downloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n",
      "Collecting pooch\n",
      "  Downloading pooch-1.7.0-py3-none-any.whl (60 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m60.9/60.9 kB\u001b[0m \u001b[31m6.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting biopython>=1.80\n",
      "  Downloading biopython-1.81-cp39-cp39-macosx_10_9_x86_64.whl (2.7 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.7/2.7 MB\u001b[0m \u001b[31m2.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0mm\n",
      "\u001b[?25hRequirement already satisfied: pandas in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from Bio) (1.4.4)\n",
      "Requirement already satisfied: tqdm in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from Bio) (4.64.1)\n",
      "Collecting mygene\n",
      "  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
      "Requirement already satisfied: numpy in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from biopython>=1.80->Bio) (1.21.5)\n",
      "Collecting biothings-client>=0.2.6\n",
      "  Downloading biothings_client-0.3.0-py2.py3-none-any.whl (29 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from pandas->Bio) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from pandas->Bio) (2022.1)\n",
      "Requirement already satisfied: platformdirs>=2.5.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from pooch->Bio) (2.5.2)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from pooch->Bio) (21.3)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests->Bio) (3.3)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests->Bio) (1.26.11)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests->Bio) (2022.9.24)\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests->Bio) (2.0.4)\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from packaging>=20.0->pooch->Bio) (3.0.9)\n",
      "Requirement already satisfied: six>=1.5 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from python-dateutil>=2.8.1->pandas->Bio) (1.16.0)\n",
      "Installing collected packages: biopython, pooch, gprofiler-official, biothings-client, mygene, Bio\n",
      "Successfully installed Bio-1.5.8 biopython-1.81 biothings-client-0.3.0 gprofiler-official-1.0.0 mygene-3.2.2 pooch-1.7.0\n"
     ]
    }
   ],
   "source": [
    "!pip install Bio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "74e368bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting modlamp\n",
      "  Downloading modlamp-4.3.0.tar.gz (8.0 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m8.0/8.0 MB\u001b[0m \u001b[31m2.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: setuptools>=20.2.2 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (63.4.1)\n",
      "Requirement already satisfied: nose>=1.3.7 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.3.7)\n",
      "Requirement already satisfied: numpy>=1.10.4 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.21.5)\n",
      "Requirement already satisfied: scipy>=0.17.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.9.1)\n",
      "Requirement already satisfied: matplotlib>=1.5.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (3.5.2)\n",
      "Requirement already satisfied: scikit-learn>=0.18.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.0.2)\n",
      "Requirement already satisfied: pandas>=0.18.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.4.4)\n",
      "Requirement already satisfied: requests>=2.11.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (2.28.1)\n",
      "Requirement already satisfied: lxml>=3.6.4 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (4.9.1)\n",
      "Requirement already satisfied: joblib>=0.15.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.1.0)\n",
      "Collecting mysql-connector-python==8.0.17\n",
      "  Downloading mysql_connector_python-8.0.17-py2.py3-none-any.whl (345 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m345.2/345.2 kB\u001b[0m \u001b[31m26.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hCollecting protobuf>=3.0.0\n",
      "  Downloading protobuf-4.22.1-cp37-abi3-macosx_10_9_universal2.whl (397 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m397.2/397.2 kB\u001b[0m \u001b[31m4.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: python-dateutil>=2.7 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (2.8.2)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (21.3)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (9.2.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (4.25.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (1.4.2)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (3.0.9)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (0.11.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from pandas>=0.18.1->modlamp) (2022.1)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (1.26.11)\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (2.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (2022.9.24)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (3.3)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from scikit-learn>=0.18.0->modlamp) (2.2.0)\n",
      "Requirement already satisfied: six>=1.5 in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (from python-dateutil>=2.7->matplotlib>=1.5.1->modlamp) (1.16.0)\n",
      "Building wheels for collected packages: modlamp\n",
      "  Building wheel for modlamp (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for modlamp: filename=modlamp-4.3.0-py3-none-any.whl size=173889 sha256=6adcff4fac8fc660d9f282b12ad660acd9c24ea8d6cf8e43c64c749542d8eeeb\n",
      "  Stored in directory: /Users/ggcl7/Library/Caches/pip/wheels/fc/09/85/aefb43b4f79b6323733e8276c157c4851409b65f77d7248ace\n",
      "Successfully built modlamp\n",
      "Installing collected packages: protobuf, mysql-connector-python, modlamp\n",
      "Successfully installed modlamp-4.3.0 mysql-connector-python-8.0.17 protobuf-4.22.1\n"
     ]
    }
   ],
   "source": [
    "!pip install modlamp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9c40e3ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: tqdm in /Users/ggcl7/opt/anaconda3/lib/python3.9/site-packages (4.64.1)\r\n"
     ]
    }
   ],
   "source": [
    "!pip install tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5e073c2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis as PA\n",
    "from modlamp.descriptors import PeptideDescriptor, GlobalDescriptor\n",
    "from sklearn.model_selection import train_test_split\n",
    "import pandas as pd\n",
    "import os, re, math, platform\n",
    "from tqdm import tqdm\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "from Bio import Seq, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "76593489",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:469: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:471: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:473: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:469: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:471: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:473: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "/var/folders/19/51xjwpd57m1g5zzk1p6ps9g40000gn/T/ipykernel_5123/4061532156.py:469: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "  if n is 1:\n",
      "/var/folders/19/51xjwpd57m1g5zzk1p6ps9g40000gn/T/ipykernel_5123/4061532156.py:471: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "  elif n is 2:\n",
      "/var/folders/19/51xjwpd57m1g5zzk1p6ps9g40000gn/T/ipykernel_5123/4061532156.py:473: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "  elif n is 3:\n"
     ]
    }
   ],
   "source": [
    "# Miscellaneous\n",
    "_AALetter = ['A', 'C', 'D', 'E', 'F', 'G', 'H',\n",
    "             'I', 'K', 'L', 'M', 'N', 'P', 'Q',\n",
    "             'R', 'S', 'T', 'V', 'W', 'Y']\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "n_gram statistics\n",
    "\"\"\"\n",
    "\n",
    "_AALetter = ['A', 'C', 'D', 'E', 'F', 'G', 'H', \n",
    "            'I', 'K', 'L', 'M', 'N', 'P', 'Q', \n",
    "            'R', 'S', 'T', 'V', 'W', 'Y']\n",
    "\n",
    "def get_aan_corpus(n=2):\n",
    "    '''\n",
    "    Get AA corpus of n_gram (e.g. Di, Tri, etc.)\n",
    "    Output: AA n_gram corpus ((e.g. Di:400, Tri:3000, etc.))\n",
    "    '''\n",
    "    n_corpus = []\n",
    "    if n <= 2:\n",
    "        for i in _AALetter:\n",
    "            for j in _AALetter:\n",
    "               n_corpus.append(\"{}{}\".format(i, j))\n",
    "        return n_corpus\n",
    "    for i in get_aan_corpus(n - 1):\n",
    "        for j in _AALetter:\n",
    "            n_corpus.append(\"{}{}\".format(i, j))\n",
    "    return n_corpus\n",
    "\n",
    "\n",
    "def get_ngram_counts(seq, n=2):\n",
    "    '''\n",
    "    Get n_gram statistics\n",
    "    Input: peptide sequence and n\n",
    "    Ouput: n_gram statistic (dictionary) {A.A Corp: Counts}\n",
    "    '''\n",
    "    # Get the name of ngram feature\n",
    "    if n == 2:\n",
    "        prefix = 'DPC'\n",
    "    elif n == 3:\n",
    "        prefix = 'TPC'\n",
    "    else:\n",
    "        prefix = '{}gram'.format(n)\n",
    "\n",
    "    ngrams = [seq[i: i + n] for i in range(len(seq) - n + 1)]\n",
    "    n_corpus = get_aan_corpus(n)\n",
    "    ngram_stat = {}\n",
    "    for aa_ng in n_corpus:\n",
    "        ngram_stat['{}|{}'.format(prefix, aa_ng)] = ngrams.count(aa_ng) / len(ngrams) * 100\n",
    "    return ngram_stat\n",
    "\n",
    "\n",
    "def minSequenceLength(fastas):\n",
    "    minLen = 10000\n",
    "    for i in fastas:\n",
    "        if minLen > len(i[1]):\n",
    "            minLen = len(i[1])\n",
    "    return minLen\n",
    "\n",
    "\n",
    "def minSequenceLengthWithNormalAA(fastas):\n",
    "    minLen = 10000\n",
    "    for i in fastas:\n",
    "        if minLen > len(re.sub('-', '', i[1])):\n",
    "            minLen = len(re.sub('-', '', i[1]))\n",
    "    return minLen\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "    input.fasta:      the input protein sequence file in fasta format.\n",
    "    k_space:          the gap of two amino acids, integer, defaule: 5\n",
    "    output:           the encoding file, default: 'encodings.tsv'\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "def generateGroupPairs(groupKey):\n",
    "    gPair = {}\n",
    "    for key1 in groupKey:\n",
    "        for key2 in groupKey:\n",
    "            gPair['CKSAAGP|'+key1+'.'+key2] = 0\n",
    "    return gPair\n",
    "\n",
    "\n",
    "\n",
    "def Rvalue(aa1, aa2, AADict, Matrix):\n",
    "    return sum([(Matrix[i][AADict[aa1]] - Matrix[i][AADict[aa2]]) ** 2 for i in range(len(Matrix))]) / len(Matrix)\n",
    "\n",
    "\n",
    "def paac(fastas, lambdaValue=30, w=0.05, **kw):\n",
    "    if minSequenceLengthWithNormalAA(fastas) < lambdaValue + 1:\n",
    "        print('Error: all the sequence length should be larger than the lambdaValue+1: ' + str(lambdaValue + 1) + '\\n\\n')\n",
    "        return 0\n",
    "\n",
    "    dataFile = re.sub('codes$', '', os.path.split(os.path.realpath('__file__'))[0]) + r'\\data\\PAAC.txt' if platform.system() == 'Windows' else re.sub('codes$', '', os.path.split(os.path.realpath('__file__'))[0]) + '/data/PAAC.txt'\n",
    "    with open(dataFile) as f:\n",
    "        records = f.readlines()\n",
    "    AA = ''.join(records[0].rstrip().split()[1:])\n",
    "    AADict = {}\n",
    "    for i in range(len(AA)):\n",
    "        AADict[AA[i]] = i\n",
    "    AAProperty = []\n",
    "    AAPropertyNames = []\n",
    "    for i in range(1, len(records)):\n",
    "        array = records[i].rstrip().split() if records[i].rstrip() != '' else None\n",
    "        AAProperty.append([float(j) for j in array[1:]])\n",
    "        AAPropertyNames.append(array[0])\n",
    "\n",
    "    AAProperty1 = []\n",
    "    for i in AAProperty:\n",
    "        meanI = sum(i) / 20\n",
    "        fenmu = math.sqrt(sum([(j-meanI)**2 for j in i])/20)\n",
    "        AAProperty1.append([(j-meanI)/fenmu for j in i])\n",
    "\n",
    "    encodings = []\n",
    "    header = ['#']\n",
    "    for aa in AA:\n",
    "        header.append('PAAC|' + aa)\n",
    "    for n in range(1, lambdaValue + 1):\n",
    "        header.append('PAAC|lambda' + str(n))\n",
    "    encodings.append(header)\n",
    "\n",
    "    for i in fastas:\n",
    "        name, sequence = i[0], re.sub('-', '', i[1])\n",
    "        code = [name]\n",
    "        theta = []\n",
    "        for n in range(1, lambdaValue + 1):\n",
    "            theta.append(\n",
    "                sum([Rvalue(sequence[j], sequence[j + n], AADict, AAProperty1) for j in range(len(sequence) - n)]) / (\n",
    "                len(sequence) - n))\n",
    "        myDict = {}\n",
    "        for aa in AA:\n",
    "            myDict[aa] = sequence.count(aa)\n",
    "        code = code + [myDict[aa] / (1 + w * sum(theta)) for aa in AA]\n",
    "        code = code + [(w * j) / (1 + w * sum(theta)) for j in theta]\n",
    "        encodings.append(code)\n",
    "    return encodings\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def insert_phycs(seq_df):\n",
    "    seq_df = seq_df.copy()\n",
    "    #  Function for compute Isoelectric Point or net_charge of peptide\n",
    "    def get_ieq_nc(seq, is_iep=True):\n",
    "        protparam = PA(seq)\n",
    "        return protparam.isoelectric_point() if is_iep else protparam.charge_at_pH(7.0)\n",
    "\n",
    "    # Calculating IsoElectricPoints and NeutralCharge\n",
    "    data_size = seq_df.size\n",
    "    seq_df['PHYC|IEP'] = list(map(get_ieq_nc, seq_df['Sequence'], [True] * data_size))  # IsoElectricPoints\n",
    "    seq_df['PHYC|Net Charge'] = list(map(get_ieq_nc, seq_df['Sequence'], [False] * data_size))  # Charge(Neutral)\n",
    "\n",
    "    # Calculating hydrophobic moment (My assume all peptides are alpha-helix)\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'eisenberg')\n",
    "    descrpt.calculate_moment(window=1000, angle=100, modality='max')\n",
    "    seq_df['PHYC|Hydrophobic Moment'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating \"Hopp-Woods\" hydrophobicity\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'hopp-woods')\n",
    "    descrpt.calculate_global()\n",
    "    seq_df['PHYC|Hydrophobicity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Energy of Transmembrane Propensity\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'tm_tend')\n",
    "    descrpt.calculate_global()\n",
    "    seq_df['PHYC|Transmembrane Propensity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Aromaticity\n",
    "    descrpt = GlobalDescriptor(seq_df['Sequence'].values)\n",
    "    descrpt.aromaticity()\n",
    "    seq_df['PHYC|Aromacity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Levitt_alpha_helical Propensity\n",
    "    descrpt = PeptideDescriptor(seq_df['Sequence'].values, 'levitt_alpha')\n",
    "    descrpt.calculate_global()\n",
    "    seq_df['PHYC|Alpha Helical Propensity'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Aliphatic Index\n",
    "    descrpt = GlobalDescriptor(seq_df['Sequence'].values)\n",
    "    descrpt.aliphatic_index()\n",
    "    seq_df['PHYC|Aliphatic Index'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    # Calculating Boman Index\n",
    "    descrpt = GlobalDescriptor(seq_df['Sequence'].values)\n",
    "    descrpt.boman_index()\n",
    "    seq_df['PHYC|Boman Index'] = descrpt.descriptor.reshape(-1)\n",
    "\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "'''\n",
    "Insert Amino acid composition to the sequence data_frame\n",
    "Input: sequence data_frame {IDx: Seq_x}\n",
    "Output: data_frame of Peptide Seq {IDx: Seq_x, ..., AAC_Ax ... AAC_Yx}\n",
    "'''\n",
    "\n",
    "\n",
    "def insert_aac(seq_df):\n",
    "    seq_df = seq_df.copy()\n",
    "    # Compute AAC for peptide in specific A.A\n",
    "    def get_aac(seq, aa):\n",
    "        return seq.count(aa) / len(seq) * 100\n",
    "\n",
    "    # processing data_frame\n",
    "    data_size = seq_df.size\n",
    "    for ll in _AALetter:\n",
    "        seq_df['AAC|{}'.format(ll)] = list(map(get_aac, seq_df['Sequence'], [ll] * data_size))\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def insert_paac(seq_df, lamb=3, w=0.1):\n",
    "    seq_df = seq_df.copy()\n",
    "    fastas = [[idx, seq] for idx, seq in zip(seq_df['Id'], seq_df['Sequence'])]\n",
    "    encoding = paac(fastas, lambdaValue=lamb, w=w)\n",
    "    encoding = pd.DataFrame(encoding[1:], columns=encoding[0])\n",
    "    seq_df = pd.concat([seq_df, encoding.iloc[:, 1:]], axis=1)\n",
    "    return seq_df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def construct_features(seq_df, paaclamb=3, paacw=0.5):\n",
    "    \"\"\"\n",
    "    Construct Features for the AVPIden. We first investigated physiochemical feautres, AAC features, DiC features, \n",
    "    CKSAAGP features, PAAC features, and PHYC features.\n",
    "    Parameters are pre-set according to the sequence identities.\n",
    "    \"\"\"\n",
    "    seq_df = insert_aac(seq_df)\n",
    "    seq_df = insert_paac(seq_df, lamb=paaclamb, w=paacw)\n",
    "    seq_df = insert_phycs(seq_df)\n",
    "\n",
    "    return seq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a2d488e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_fasta(df, file_path, abbr_columns=None):\n",
    "    \"\"\"\n",
    "    Save dataframe to a .fasta file, the df should contain at least columns named \"Id\" and \"Sequence\"\n",
    "    \n",
    "    df: dataframe for saving .fasta\n",
    "    file_path: path(string) for saving the fasta file\n",
    "    abbr_columns: string columns for adding abbreviations. Multiple abbr are splited by '|'.\n",
    "    \"\"\"\n",
    "    Seqrecords = [SeqIO.SeqRecord(id=row['Id'], \n",
    "                              seq=Seq.Seq(row['Sequence']), \n",
    "                              description='|'.join(row[abbr_columns] if abbr_columns is not None else \"\")) \\\n",
    "             for idn, row in df.iterrows()]\n",
    "    with open(file_path, 'w+') as fhandle:\n",
    "        SeqIO.write(Seqrecords, fhandle, \"fasta-2line\")\n",
    "        print(\"Saved {:d} sequences.\".format(len(Seqrecords)))\n",
    "\n",
    "\n",
    "def read_fasta(file_path):\n",
    "    f = open(file_path, 'r', encoding='utf-8')\n",
    "    fasta_list = np.array(f.readlines())\n",
    "    # print(fasta_list)\n",
    "    new_fasta = []\n",
    "    for flag in range(0, len(fasta_list), 2):\n",
    "        fasta_str = [fasta_list[flag].strip('\\n').strip(), fasta_list[flag + 1].strip('\\n').strip()]\n",
    "        new_fasta.append(fasta_str)\n",
    "    # print(new_fasta)\n",
    "    seq_df = pd.DataFrame(data=new_fasta, columns=['Id', 'Sequence'])\n",
    "    # print(new_fasta1.shape)\n",
    "    return seq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3edcc430",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_fasta = read_fasta(\"non_THP.txt\")\n",
    "seq_feature = construct_features(seq_fasta)\n",
    "seq_feature.to_csv(\"non_THP_AAC_PAAC_PHYC.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "24470880",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>AAC|A</th>\n",
       "      <th>AAC|C</th>\n",
       "      <th>AAC|D</th>\n",
       "      <th>AAC|E</th>\n",
       "      <th>AAC|F</th>\n",
       "      <th>AAC|G</th>\n",
       "      <th>AAC|H</th>\n",
       "      <th>AAC|I</th>\n",
       "      <th>...</th>\n",
       "      <th>PAAC|lambda3</th>\n",
       "      <th>PHYC|IEP</th>\n",
       "      <th>PHYC|Net Charge</th>\n",
       "      <th>PHYC|Hydrophobic Moment</th>\n",
       "      <th>PHYC|Hydrophobicity</th>\n",
       "      <th>PHYC|Transmembrane Propensity</th>\n",
       "      <th>PHYC|Aromacity</th>\n",
       "      <th>PHYC|Alpha Helical Propensity</th>\n",
       "      <th>PHYC|Aliphatic Index</th>\n",
       "      <th>PHYC|Boman Index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;ACP203</td>\n",
       "      <td>KAKLF</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.327496</td>\n",
       "      <td>10.002737</td>\n",
       "      <td>1.758104</td>\n",
       "      <td>0.316465</td>\n",
       "      <td>0.240000</td>\n",
       "      <td>-0.548000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>1.224000</td>\n",
       "      <td>98.000000</td>\n",
       "      <td>0.278000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;ACP217</td>\n",
       "      <td>ACSAG</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.138602</td>\n",
       "      <td>5.561548</td>\n",
       "      <td>-0.214026</td>\n",
       "      <td>0.284292</td>\n",
       "      <td>-0.340000</td>\n",
       "      <td>-0.052000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.014000</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>-0.488000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;ACP259</td>\n",
       "      <td>WALAL</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.404608</td>\n",
       "      <td>5.525000</td>\n",
       "      <td>-0.239898</td>\n",
       "      <td>0.183105</td>\n",
       "      <td>-1.600000</td>\n",
       "      <td>1.186000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>1.234000</td>\n",
       "      <td>196.000000</td>\n",
       "      <td>-3.158000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;ACP279</td>\n",
       "      <td>FAKLA</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.027398</td>\n",
       "      <td>8.750052</td>\n",
       "      <td>0.759103</td>\n",
       "      <td>0.714502</td>\n",
       "      <td>-0.460000</td>\n",
       "      <td>0.220000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>1.236000</td>\n",
       "      <td>118.000000</td>\n",
       "      <td>-1.194000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;ACP293</td>\n",
       "      <td>PPPEE</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.304677</td>\n",
       "      <td>4.240666</td>\n",
       "      <td>-2.030613</td>\n",
       "      <td>0.196176</td>\n",
       "      <td>1.200000</td>\n",
       "      <td>-2.024000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.888000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.724000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>&gt;AMP270</td>\n",
       "      <td>RIGVLLARLPKLFSLFKLMGKKV</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.695652</td>\n",
       "      <td>8.695652</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>...</td>\n",
       "      <td>0.197175</td>\n",
       "      <td>11.999968</td>\n",
       "      <td>5.756086</td>\n",
       "      <td>0.339831</td>\n",
       "      <td>-0.178261</td>\n",
       "      <td>0.009565</td>\n",
       "      <td>0.086957</td>\n",
       "      <td>1.077826</td>\n",
       "      <td>148.260870</td>\n",
       "      <td>0.040000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>&gt;AMP341</td>\n",
       "      <td>LLDCWVRLGRYLLRRLKTPFTRL</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.219064</td>\n",
       "      <td>11.437223</td>\n",
       "      <td>4.749273</td>\n",
       "      <td>0.694884</td>\n",
       "      <td>-0.134783</td>\n",
       "      <td>-0.171739</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>1.030000</td>\n",
       "      <td>131.304348</td>\n",
       "      <td>2.093043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>&gt;AMP355</td>\n",
       "      <td>SNMIEGVFAKGFKKASHLFKGIG</td>\n",
       "      <td>8.695652</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>13.043478</td>\n",
       "      <td>17.391304</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>8.695652</td>\n",
       "      <td>...</td>\n",
       "      <td>0.225493</td>\n",
       "      <td>10.000416</td>\n",
       "      <td>2.546135</td>\n",
       "      <td>0.458631</td>\n",
       "      <td>-0.060870</td>\n",
       "      <td>-0.273913</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>1.033478</td>\n",
       "      <td>72.173913</td>\n",
       "      <td>0.419130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>&gt;AMP408</td>\n",
       "      <td>FALALKALKKALKKLKKALKKAL</td>\n",
       "      <td>26.086957</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.190572</td>\n",
       "      <td>10.903296</td>\n",
       "      <td>8.751111</td>\n",
       "      <td>0.639309</td>\n",
       "      <td>0.386957</td>\n",
       "      <td>-0.614783</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>1.260000</td>\n",
       "      <td>144.782609</td>\n",
       "      <td>0.072609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>&gt;AMP427</td>\n",
       "      <td>KQRNRWEWRPDFKSKKVKISLPC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.347826</td>\n",
       "      <td>...</td>\n",
       "      <td>0.331065</td>\n",
       "      <td>10.561033</td>\n",
       "      <td>5.749107</td>\n",
       "      <td>0.086677</td>\n",
       "      <td>0.678261</td>\n",
       "      <td>-1.243043</td>\n",
       "      <td>0.130435</td>\n",
       "      <td>1.030435</td>\n",
       "      <td>46.521739</td>\n",
       "      <td>3.661739</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>651 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Id                 Sequence      AAC|A      AAC|C     AAC|D  \\\n",
       "0    >ACP203                    KAKLF  20.000000   0.000000  0.000000   \n",
       "1    >ACP217                    ACSAG  40.000000  20.000000  0.000000   \n",
       "2    >ACP259                    WALAL  40.000000   0.000000  0.000000   \n",
       "3    >ACP279                    FAKLA  40.000000   0.000000  0.000000   \n",
       "4    >ACP293                    PPPEE   0.000000   0.000000  0.000000   \n",
       "..       ...                      ...        ...        ...       ...   \n",
       "646  >AMP270  RIGVLLARLPKLFSLFKLMGKKV   4.347826   0.000000  0.000000   \n",
       "647  >AMP341  LLDCWVRLGRYLLRRLKTPFTRL   0.000000   4.347826  4.347826   \n",
       "648  >AMP355  SNMIEGVFAKGFKKASHLFKGIG   8.695652   0.000000  0.000000   \n",
       "649  >AMP408  FALALKALKKALKKLKKALKKAL  26.086957   0.000000  0.000000   \n",
       "650  >AMP427  KQRNRWEWRPDFKSKKVKISLPC   0.000000   4.347826  4.347826   \n",
       "\n",
       "         AAC|E      AAC|F      AAC|G     AAC|H     AAC|I  ...  PAAC|lambda3  \\\n",
       "0     0.000000  20.000000   0.000000  0.000000  0.000000  ...      0.327496   \n",
       "1     0.000000   0.000000  20.000000  0.000000  0.000000  ...      0.138602   \n",
       "2     0.000000   0.000000   0.000000  0.000000  0.000000  ...      0.404608   \n",
       "3     0.000000  20.000000   0.000000  0.000000  0.000000  ...      0.027398   \n",
       "4    40.000000   0.000000   0.000000  0.000000  0.000000  ...      0.304677   \n",
       "..         ...        ...        ...       ...       ...  ...           ...   \n",
       "646   0.000000   8.695652   8.695652  0.000000  4.347826  ...      0.197175   \n",
       "647   0.000000   4.347826   4.347826  0.000000  0.000000  ...      0.219064   \n",
       "648   4.347826  13.043478  17.391304  4.347826  8.695652  ...      0.225493   \n",
       "649   0.000000   4.347826   0.000000  0.000000  0.000000  ...      0.190572   \n",
       "650   4.347826   4.347826   0.000000  0.000000  4.347826  ...      0.331065   \n",
       "\n",
       "      PHYC|IEP  PHYC|Net Charge  PHYC|Hydrophobic Moment  PHYC|Hydrophobicity  \\\n",
       "0    10.002737         1.758104                 0.316465             0.240000   \n",
       "1     5.561548        -0.214026                 0.284292            -0.340000   \n",
       "2     5.525000        -0.239898                 0.183105            -1.600000   \n",
       "3     8.750052         0.759103                 0.714502            -0.460000   \n",
       "4     4.240666        -2.030613                 0.196176             1.200000   \n",
       "..         ...              ...                      ...                  ...   \n",
       "646  11.999968         5.756086                 0.339831            -0.178261   \n",
       "647  11.437223         4.749273                 0.694884            -0.134783   \n",
       "648  10.000416         2.546135                 0.458631            -0.060870   \n",
       "649  10.903296         8.751111                 0.639309             0.386957   \n",
       "650  10.561033         5.749107                 0.086677             0.678261   \n",
       "\n",
       "     PHYC|Transmembrane Propensity  PHYC|Aromacity  \\\n",
       "0                        -0.548000        0.200000   \n",
       "1                        -0.052000        0.000000   \n",
       "2                         1.186000        0.200000   \n",
       "3                         0.220000        0.200000   \n",
       "4                        -2.024000        0.000000   \n",
       "..                             ...             ...   \n",
       "646                       0.009565        0.086957   \n",
       "647                      -0.171739        0.130435   \n",
       "648                      -0.273913        0.130435   \n",
       "649                      -0.614783        0.043478   \n",
       "650                      -1.243043        0.130435   \n",
       "\n",
       "     PHYC|Alpha Helical Propensity  PHYC|Aliphatic Index  PHYC|Boman Index  \n",
       "0                         1.224000             98.000000          0.278000  \n",
       "1                         1.014000             40.000000         -0.488000  \n",
       "2                         1.234000            196.000000         -3.158000  \n",
       "3                         1.236000            118.000000         -1.194000  \n",
       "4                         0.888000              0.000000          2.724000  \n",
       "..                             ...                   ...               ...  \n",
       "646                       1.077826            148.260870          0.040000  \n",
       "647                       1.030000            131.304348          2.093043  \n",
       "648                       1.033478             72.173913          0.419130  \n",
       "649                       1.260000            144.782609          0.072609  \n",
       "650                       1.030435             46.521739          3.661739  \n",
       "\n",
       "[651 rows x 54 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0da4ddc9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5f28c7b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
